### PR TITLE
Pbi 1648

### PR DESCRIPTION
**Pull request by pilievOoyala**
Taking out preloading of vast ads. Ads now load only when they are about to play, using a placeholder linear-overlay in the timeline(similar to what we do in freewheel).  And unit tests were updated to reflect the changes.

### CODE REVIEW

File: js/ad_manager_vast.js [hunk 1]
```diff
@@ -739,34 +731,27 @@ OO.Ads.manager(function(_, $) {
...
-            loadedAds = true;
-          }
-        }
+      var ad = amcAd.ad;
+      this.amc.notifyPodStarted(amcAd.id, 1);
+      if (this.adURLOverride) {
+        override = true;
+        ad.tag_url = this.adURLOverride;
+      }
+      var time = typeof ad.time !== 'undefined' ? ad.time : ad.position;
```
Comment (aeng7) [inline js/ad_manager_vast.js]: Is var time used for anything?

File: js/ad_manager_vast.js [hunk 2]
```diff
@@ -867,8 +877,18 @@ OO.Ads.manager(function(_, $) {
...
-      currentAd = adWrapper;
+      if (adWrapper) {
+        currentAd = adWrapper;
+        if (currentAd.isAdRequest) {
+          loadAd(currentAd);
+        } else {
+          playLoadedAd(adWrapper);
+        }
+      }
+    }
```
Comment (aeng7) [inline js/ad_manager_vast.js]: Missing semicolon

File: js/ad_manager_vast.js (hunk 1 above)
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: not anymore, good catch.

File: js/ad_manager_vast.js [hunk 3]
```diff
@@ -867,8 +877,18 @@ OO.Ads.manager(function(_, $) {
...
 
+    var playLoadedAd = _.bind(function(adWrapper) {
```
Comment (aeng7) [inline js/ad_manager_vast.js]: The vast ad manager uses underscore notation for private members. Can we change this to _playLoadedAd. Also, can we add a jsdoc for this?

File: js/ad_manager_vast.js [hunk 4]
```diff
@@ -801,7 +786,32 @@ OO.Ads.manager(function(_, $) {
...
+          };
+
+          if (ad.position_type == 't') {
+            adData.position = ad.time/1000;
+          } else if (ad.position_type == 'p') {
+            //TODO
+            //adData.position =
+          }
+          var amcAd = new this.amc.Ad(adData);
+          amcAd.isAdRequest = true;
```
Comment (aeng7) [inline js/ad_manager_vast.js]: The IMA ad manager uses an ad request type "adRequest" to mark an ad as an ad request. We can do the same thing here to maintain consistency across ad managers.

File: js/ad_manager_vast.js [hunk 5]
```diff
@@ -1921,6 +1941,13 @@ OO.Ads.manager(function(_, $) {
       } else {
         failedAd();
       }
+
+      if (currentAd && currentAd.isAdRequest)
```
Comment (aeng7) [inline js/ad_manager_vast.js]: Is an ad request ad able to make it to the handleAds function? handleAds is only called after we receive the Vast response in the onVastResponse function.

File: test/unit-test-helpers/mock_amc.js [hunk 6]
```diff
@@ -42,7 +42,8 @@ fake_amc = function() {
     playerSkinPluginsElement : $("<div>", {}),
     adWrapper : $("<div>", {}),
     transitionToAd: function(){},
-    transitionToMainContent: function(){}
+    transitionToMainContent: function(){},
+    createAdVideoElement: function() {}
```
Comment (aeng7) [inline test/unit-test-helpers/mock_amc.js]: There was a recent change that removed the calling of createAdVideoElement by the vast ad manager. Angie and I had a discussion and agreed this should not be an API for the ad plugins to use. Can you try removing this? If this breaks some unit tests, we can merge this branch with master prior to pushing.

File: test/unit-tests/vast_test.js [hunk 7]
```diff
@@ -120,13 +119,14 @@ describe('ad_manager_vast', function() {
...
-    initalPlay();
+    vastAdManager.loadMetadata(server, {}, content);
+    initialPlay();
+    vastAdManager.initialPlay();
     xml = xml || vpaidLinearXML;
     vastAdManager.onVastResponse(preroll, xml);
   };
 
-  var initalPlay = function() {
+  var initialPlay = function() {
```
Comment (aeng7) [inline test/unit-tests/vast_test.js]: Nice catch! :D

File: test/unit-tests/vast_test.js [hunk 8]
```diff
@@ -443,28 +383,31 @@ describe('ad_manager_vast', function() {
...
+      "html5_ad_server": "http://blah"}, {}, content);
+    expect(amc.timeline.length).to.be(1);
+
+    initialPlay();
+    vastAdManager.initialPlay();
     vastAdManager.onVastResponse(vast_ad,'<VAST></VAST>');
+
     expect(errorType.length > 0).to.be(true);
-    expect(amc.timeline.length).to.be(0);
+    expect(amc.timeline.length).to.be(1);
```
Comment (aeng7) [inline test/unit-tests/vast_test.js]: Do we check the 2nd element in the array for the unit tests because the first element is the ad request ad?

Comment (pilievOoyala) [inline test/unit-tests/vast_test.js]: Yes. I'd like to revamp these tests but we should do the moving of the AMC into this repo first.  There is too much emulation going on.

Comment (aeng7) [inline test/unit-tests/vast_test.js]: Sounds good.

**Comment by aeng7** [comment]
LGTM. Waiting on the mjolnir side.

**Pull request merged**